metatensor-torch/src/atomistic/model.cpp

```diff
@@ -731,3 +759,117 @@ torch::jit::Module metatensor_torch::load_atomistic_model(
     check_atomistic_model(path);
     return torch::jit::load(path, device);
 }
+
+/******************************************************************************/
+/******************************************************************************/
+
+/// normalize a string to lower case without any spaces
+static std::string normalize_string(std::string value) {
+    std::transform(value.begin(), value.end(), value.begin(),
+        [](unsigned char c){ return std::tolower(c); }
+    );
+
+    // remove all spaces in e.g. `kcal /   mol`.
+    auto new_end = std::remove_if(value.begin(), value.end(),
+        [](unsigned char c){ return std::isspace(c); }
+    );
+    value.erase(new_end, value.end());
+    return value;
+}
+
+/// Information for unit conversion for this physical quantity
+struct Quantity {
+    /// the quantity name
+    std::string name;
+
+    /// baseline unit for this quantity
+    std::string baseline;
+    /// set of conversion from the key to the baseline unit
+    std::unordered_map<std::string, double> conversions;
+
+    std::string normalize_unit(const std::string& original_unit) {
+        if (original_unit.empty()) {
+            return original_unit;
+        }
+
+        std::string unit = normalize_string(original_unit);
+        if (this->conversions.find(unit) == this->conversions.end()) {
+            auto valid_units = std::vector<std::string>();
+            for (const auto& it: this->conversions) {
+                valid_units.emplace_back(it.first);
+            }
+
+            C10_THROW_ERROR(ValueError,
+                "unknown unit '" + original_unit + "' for " + name + ", "
+                "only [" + torch::str(valid_units) + "] are supported"
+            );
+        }
+
+        return unit;
+    }
+
+    double conversion(const std::string& from_unit, const std::string& to_unit) {
+        auto from = this->normalize_unit(from_unit);
+        auto to = this->normalize_unit(to_unit);
+
+        if (from.empty() || to.empty()) {
+            return 1.0;
+        }
+
+        return this->conversions.at(to) / this->conversions.at(from);
+    }
+};
+
+static std::unordered_map<std::string, Quantity> KNOWN_QUANTITIES = {
+    {"length", Quantity{/* name */ "length", /* baseline */ "angstrom", {
+        {"angstrom", 1.0},
+        {"bohr", 1.8897261258369282},
+        {"nanometer", 0.1},
+        {"nm", 0.1},
+    }}},
+    {"energy", Quantity{/* name */ "energy", /* baseline */ "ev", {
+        {"ev", 1.0},
+        {"mev", 1000.0},
+        {"hartree", 0.03674932247495664},
+        {"kcal/mol", 23.060548012069496},
+        {"kj/mol", 96.48533288249877},
+    }}},
+};
+
+void metatensor_torch::validate_quantity(const std::string& quantity) {
```

Comment:
Am I reading this correctly that either a quantity is known or empty string, otherwise we throw an error? This means if I am doing `"dipole"` is not suppported anymore. I like it like this.

We should adjust the docstring of `metatensor.torch.atomistic.ModelOutput` for quantity and point to a list with supported quantities and units. Maybe create a page at `docs/src/atomistic` ?

Reply:
yes, that's how it works for now. I'm not sure I love it, since it might prevent people to add units to their models if they are not known (and they might not come back & add units later). Not sure what the best solution is here. Maybe only sending a warning for unknown quantity (but error on unknown unit) would be better?

> point to a list with supported quantities and units.

It is documented in `unit_conversion_factor`: https://metatensor--533.org.readthedocs.build/en/533/atomistic/reference/models/index.html#metatensor.torch.atomistic.unit_conversion_factor

I could move it out of the function and link to it from `ModelOutput` and friends.

Reply:
> Maybe only sending a warning for unknown quantity (but error on unknown unit) would be better?

Yes this is what I had in mind when I did the checks for `metatensor-models`

> It is documented in unit_conversion_factor: https://metatensor--533.org.readthedocs.build/en/533/atomistic/reference/models/index.html#metatensor.torch.atomistic.unit_conversion_factor


Ahh I found it. Maybe moving it out in propper clean table is better. I think this is quite important that we document what quantities we support.

Reply:
Done!